pkg/boot/universalpayload/utilities.go

```diff
@@ -715,6 +766,221 @@ func constructReservedMemoryNode(rsdpBase uint64) *dt.Node {
+					}
+				}
+			}
+		}
+
+		return nil
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	return &ResourceRegions{
+		Mmio64Base:  mmio64Base,
+		Mmio64Limit: mmio64End - mmio64Base + 1,
+		Mmio32Base:  mmio32Base,
+		Mmio32Limit: mmio32End - mmio32Base + 1,
+		IoPortBase:  ioPortBase,
+		IoPortLimit: ioPortEnd - ioPortBase + 1,
+	}, nil
+}
+
+func retrieveDeviceResources(resourcePath string) ([]ResourceInfo, error) {
+	contentBytes, err := os.ReadFile(resourcePath)
+	if err != nil {
+		return nil, err
+	}
+
+	content := string(contentBytes)
+	lines := strings.Split(content, "\n")
+	var resources []ResourceInfo
+
+	for _, line := range lines {
+		line = strings.TrimSpace(line)
+		if line == "" {
+			continue
+		}
+		parts := strings.Split(line, " ")
+
+		if len(parts) == 3 {
+			base := strings.TrimSpace(parts[0])
+			end := strings.TrimSpace(parts[1])
+			attr := strings.TrimSpace(parts[2])
+
+			if attrInt, err := strconv.ParseUint(attr, 0, 64); err != nil {
+				// Should not happen, skip this line in case it happens
+				continue
+			} else {
+				var resourceType string
+				if attrInt == PciMmio64Attr {
+					resourceType = PciMmio64Type
+				} else if attrInt == PciMmio32Attr {
+					resourceType = PciMmio32Type
+				} else if attrInt&PciIoPortAttr == PciIoPortAttr {
+					// Do not care about IRQ specific bits (BITS 0~5)
+					resourceType = PciIoPortType
+				} else {
+					continue // Skip unknown resource types
+				}
+
+				resources = append(resources, ResourceInfo{
+					Type:        resourceType,
+					BaseAddress: base,
+					EndAddress:  end,
+					Attributes:  attr,
+				})
+			}
+		}
+	}
+	return resources, nil
+}
+
+func fetchAcpiMcfgData(data []byte) ([]MCFGBaseAddressAllocation, error) {
+	var mcfgDataArray []MCFGBaseAddressAllocation
+
+	// Check if the data is long enough to contain data from offset 0x2c.
+	if len(data) <= AcpiMcfgBaseAddressAllocationOffset {
+		return nil, ErrMcfgDataLenthTooShort
+	}
+
+	// Check if the magic word is "MCFG".
+	if !bytes.Equal(data[:4], []byte(AcpiMcfgSignature)) {
+		return nil, ErrMcfgSignatureMismatch
+	}
+
+	segInfoContent := data[AcpiMcfgBaseAddressAllocationOffset:]
+
+	// Check whether content in Base Address Allocation Structure is valid
+	if len(segInfoContent)%AcpiMcfgBaseAddressAllocationLenth != 0 {
+		return nil, ErrMcfgBaseAddrAllocCorrupt
+	}
+
+	for i := 0; i < len(segInfoContent); i += AcpiMcfgBaseAddressAllocationLenth {
+		mcfgDataBytes := segInfoContent[i : i+AcpiMcfgBaseAddressAllocationLenth]
+		mcfgData := MCFGBaseAddressAllocation{}
+		reader := bytes.NewReader(mcfgDataBytes)
+
+		err := binary.Read(reader, binary.LittleEndian, &mcfgData)
+		if err != nil {
+			return nil, ErrMcfgBaseAddrAllocDecode
+		}
+
+		mcfgDataArray = append(mcfgDataArray, mcfgData)
+	}
+	return mcfgDataArray, nil
+}
+
+func constructPciRootBridgeNodes() ([]*dt.Node, error) {
+	var rbNodes []*dt.Node
+
+	highPartOfUint64 := func(val uint64) uint32 {
```

Comment:
I would consider using smaller names for such small function like high64 and low64 or something.